src/modules/commander/Commander.cpp

```diff
@@ -1070,13 +975,60 @@ Commander::handle_command(vehicle_status_s *status_local, const vehicle_command_
 		cmd_result = handle_command_motor_test(cmd);
 		break;
 
+	case vehicle_command_s::VEHICLE_CMD_PREFLIGHT_REBOOT_SHUTDOWN:
+
+		// do nothing for autopilot
+		if (((int)(cmd.param1)) == 0) {
+			answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
+
+			break;
+		}
+
+		if (shutdown_if_allowed()) {
+			bool shutdown_ret_val = PX4_ERROR;
+
+			if (((int)(cmd.param1)) == 1) {
+				answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
+				px4_usleep(200000);
+				// reboot
+				shutdown_ret_val = px4_shutdown_request(true, false);
+
+			} else if (((int)(cmd.param1)) == 2) {
+				answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
+				px4_usleep(200000);
+				// shutdown
+				shutdown_ret_val = px4_shutdown_request(false, false);
+
+			} else if (((int)(cmd.param1)) == 3) {
+				answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED, command_ack_pub);
+				px4_usleep(200000);
+				// reboot to bootloader
+				shutdown_ret_val = px4_shutdown_request(true, true);
+
+			} else {
+				answer_command(cmd, vehicle_command_s::VEHICLE_CMD_RESULT_DENIED, command_ack_pub);
+				break;
+			}
+
+			if (shutdown_ret_val) {
+				mavlink_log_critical(&mavlink_log_pub, "System does not support shutdown");
```

Comment:
```suggestion
				mavlink_log_critical(&mavlink_log_pub, "System does not support shutdown type %d", shutdown_type);
```
Since it could be a specific type that is unsupported right?